File: functions/matlab/dPrime2AFC.m
function dPrime = dPrime2AFC(correctSide, responseSide);
%dPrime = dPrime2AFC(correctSide, responseSide);
%
% Calculate the d' score for a 2AFC task.
% 
% INPUTS: -correctSide: The assigned correect side
%         -responseSide: The side chosen by the animal (NaN tolerated)
%
% OUTPUT: -dPrime: The dPrime score, currently without 1/2^(1/2)
%                  correction.
%
% LO, 5/5/2021
%--------------------------------------------------------------------------

%Input check
correctSide = correctSide(~isnan(responseSide)); %Make sure to only consider valid trials
responseSide = responseSide(~isnan(responseSide));

%Determine the side to look at
sideCodes = unique(responseSide); %To make it neutral to the input code
theSide = sideCodes(1); %First take the lower one (usually left)


hitRate = sum(correctSide == theSide & responseSide == theSide)/sum(correctSide == theSide);
%Is defined as the proportion the animal responded correctly with left side
%when left side was true (so, out of all left side is true)
falseAlarmRate = sum(correctSide ~= theSide & responseSide == theSide)/sum(correctSide ~= theSide);
%Is defined as the proportion the animal responded incorrectly with left
%side when the right side was true (so, out of all right side is true)

dPrime = norminv(hitRate) - norminv(falseAlarmRate); %Norminv is the inverse normal distribution fucntion
%dPrime = (1/sqrt(2))*(norminv(hitRate) - norminv(falseAlarmRate)); %This
%is when comapring a 2AFC to a detection task.
% 
% % Handle degenerate cases
% if dPrime == Inf || dPrime == -Inf %Check for misclassifications
%     if ~(hitRate == 1 && falseAlarmrate == 0) %If it is not the ideal observer
%       if hitrate == 1 %The 
%           dPrime =  norminv(1-falseAlaramRate);
%        
%     if (hitRate-falseAlarmRate) < 0.5 %This is only the case when animals are so biased that they only choose one side!
%         dPrime = 0; %Force 0 discrimination
%     end
% end

File: functions/python/dPrime2AFC.py
def dprime_2afc(correct_side, response_side):
  """Calculate the d' score for a 2AFC task.

  Args:
    correct_side: The assigned correct side.
    response_side: The side chosen by the animal (NaN tolerated).

  Returns:
    The d' score, currently without 1/2^(1/2) correction.
  """

  # Input check
  correct_side = correct_side[~np.isnan(response_side)]
  response_side = response_side[~np.isnan(response_side)]

  # Determine the side to look at
  side_codes = np.unique(response_side)
  the_side = side_codes[0]

  hit_rate = np.sum(correct_side == the_side & response_side == the_side) / np.sum(correct_side == the_side)
  false_alarm_rate = np.sum(correct_side != the_side & response_side == the_side) / np.sum(correct_side != the_side)

  d_prime = scipy.stats.norm.ppf(hit_rate) - scipy.stats.norm.ppf(false_alarm_rate)

  # Handle degenerate cases
  if d_prime == np.inf or d_prime == -np.inf:
    if hit_rate != 1 and false_alarm_rate != 0:
      if hit_rate == 1:
        d_prime = scipy.stats.norm.ppf(1 - false_alarm_rate)
      elif false_alarm_rate == 1:
        d_prime = scipy.stats.norm.ppf(1 - hit_rate)
      else:
        d_prime = 0
    elif hit_rate - false_alarm_rate < 0.5:
      d_prime = 0

  return d_prime
